nixos: Add meta.tests

roberth opened the pull request:
This provides the needed infrastructure to make tooling like ofborg build the relevant tests, given a set of changed files.

Having mandatory CI for modules is not just crucial for DX, but also prerequisite to test-driven expansion of NixOS use cases, such as alternate base module sets. (minimal module list and similar ideas)

The option is documented.
We can provide additional instructional docs elsewhere when this is wired into ofborg.


### Why not infer the mapping automatically?

Basically the same reasons we have `passthru.tests.nixos`.

1. Too costly. We'd have to evaluate all tests to figure out which ones changed.
2. For some modules it is not feasible to run all affected tests, because that set is too large. We need a representative smaller set.
3. This could perhaps be scripted as a one-off or infrequent process that updates the mapping, but then it is *time of all evals* * *number of files under `nixos/`... And then there's still some risk that it's not a good mapping for some reason. Doesn't solve (2)


<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Context

- Closes #148364

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [Nixpkgs 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/doc/release-notes/rl-2511.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/doc/manual/release-notes/rl-2505.section.md) Nixpkgs Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
- [NixOS 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2511.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) NixOS Release notes)
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

roberth committed:
nixos: Add meta.tests

This provides the needed infrastructure to make tooling like ofborg
build the relevant tests, given a set of changed files.

Having mandatory CI for modules is not just crucial for DX, but also
prerequisite to test-driven expansion of NixOS use cases, such as
alternate base module sets. (minimal module list and similar ideas)

MattSturgeon commented on nixos/modules/misc/meta.nix:
```diff
@@ -39,6 +39,64 @@ let
   };
+
+  /**
+    Custom type for `meta.tests` option.
```
This is kinda redundant given the name & context, however many of the other custom types in this file describe how the value is merged; e.g.

```
# Returns tuples of
#   { file = "module location"; value = <path/to/doc.xml>; }
```

and 

```
# Returns list of
#   { "module-file" = [
#        "maintainer1 <first@nixos.org>"
#        "maintainer2 <second@nixos.org>" ];
#   }
```

Something like that may be more useful here :slightly_smiling_face:

MattSturgeon commented on nixos/modules/misc/meta.nix:
```diff
@@ -39,6 +39,64 @@ let
+  */
+  testsType = lib.mkOptionType {
+    name = "tests";
+    check = value: lib.isFunction value;
```
Was this intentional?

```suggestion
    check = lib.isFunction;
```

MattSturgeon commented on nixos/modules/misc/meta.nix:
```diff
@@ -39,6 +39,64 @@ let
+            An attrset structure of <testName>.<fileName> = derivation;
+          */
+          transposed = lib.zipAttrsWith (_k: lib.mergeAttrsList) (
+            lib.mapAttrsToList (fileName: values: lib.mapAttrs (k: v: { "${fileName}" = v; }) values) relevant
```
Again, was this intentional?

```suggestion
            lib.mapAttrsToList (fileName: lib.mapAttrs (k: v: { "${fileName}" = v; })) relevant
```

Not really important, I'm just curious if you have a preference for either code style.

roberth replied:
Both are ok. I only have a preference in code that's performance critical, where eta-reduced performs better, producing fewer calls. Shorter error traces can also be nice.

MattSturgeon commented on nixos/modules/misc/meta.nix:
```diff
@@ -39,6 +39,64 @@ let
+
+        This is like `filter`, but solves the problem that `nix-build` ignores attributes with `.` in their names.
+      */
+      filterForNixBuild = paths: lib.concatMap lib.attrValues (lib.attrValues (filter paths));
```
Would this be more readable as a pipe?

```suggestion
      filterForNixBuild = paths: lib.pipe paths [
        filter
        builtins.attrValues
        (builtins.concatMap builtins.attrValues)
      ];
```

Although, more verbose so maybe not worth it.

---

Alternatively, how about using `collect isDerivation`? Maybe this would be less efficient, but it might convey the intention more clearly:

 ```suggestion
      filterForNixBuild = paths: lib.collect lib.isDerivation (filter paths);
```

roberth replied:
> but it might convey the intention more clearly:

It somewhat blurs the intent, because it's two layers and not something arbitrarily deep.
I'll add:

```nix
      # Two calls to `attrValues` because we have exactly two layers to flatten.
```

MattSturgeon commented on nixos/modules/misc/meta.nix:
```diff
@@ -81,8 +139,51 @@ in
+
+          This can be used for tooling to figure out which tests are relevant, given a set of changed files.
+        '';
+        default = { ... }: { };
```
```suggestion
        default = { ... }: { };
        defaultText = lib.literalExpression "{ ... }: { }";
```

or

```suggestion
        default = _: { };
        defaultText = lib.literalExpression "{ nixosTests }: { }";
```

roberth replied:
I'm un-documenting the default because it's niche and un-interesting.
The `description` covers it all.
That also means I won't add an `example` (not that you were suggesting to), as that would be confusing because the definitions and option value are so different.

MattSturgeon commented on nixos/modules/misc/meta.nix:
```diff
@@ -81,8 +139,51 @@ in
+
+          Definitions are tied to the location of the module, so the definition syntax is distinct from the option value format.
+
+          Example definition:
```
```suggestion
          For example, this definition:
```

MattSturgeon commented on nixos/modules/misc/meta.nix:
````diff
@@ -81,8 +139,51 @@ in
+          }
+          ```
+
+          Example option value:
````
```suggestion
          Would produce the following option value:
```

MattSturgeon commented on nixos/modules/misc/meta.nix:
````diff
@@ -81,8 +139,51 @@ in
+          ```nix
+          # module.nix
+          {
+            meta.tests = { nixosTests }:
````
Nit: nixfmt would format like this:

```suggestion
            meta.tests =
              { nixosTests }:
```

Should the example be consistent?

MattSturgeon reviewed (commented):
Nice idea, this looks useful.

A few minor comments, most of which I could address in a followup if preferred. Some of my comments are more questions than feedback.

DavHau commented on maintainers/nixos-tests-for-files.nix:
```diff
@@ -0,0 +1,25 @@
+  A more detailed structure than a list is available in the `meta.tests` option value.
+
+  Example:
+    nix-build maintainers/nixos-tests-for-files.nix --arg files '[ nixos/modules/services/databases/postgresql.nix ]'
```
@MattSturgeon Did you try running the command before making that comment?

DavHau commented on maintainers/nixos-tests-for-files.nix:
I guess computing a list in nix is too expensive (See robers explanation in the topic). `nixpkgs-review` can instead implement a feature to call that function with the list of changed files.

DavHau reviewed (approved):
> ### Why not infer the mapping automatically?
> 
> Basically the same reasons we have `passthru.tests.nixos`.
> 
>     1. Too costly. We'd have to evaluate all tests to figure out which ones changed.
> 
>     2. For some modules it is not feasible to run all affected tests, because that set is too large. We need a representative smaller set.
> 
>     3. This could perhaps be scripted as a one-off or infrequent process that updates the mapping, but then it is _time of all evals_ * *number of files under `nixos/`... And then there's still some risk that it's not a good mapping for some reason. Doesn't solve (2)

This reasoning makes sense to me.

I tested the interface and it works fine. I think it's a good idea in general.

DavHau commented on maintainers/nixos-tests-for-files.nix:
```diff
@@ -0,0 +1,25 @@
+  A more detailed structure than a list is available in the `meta.tests` option value.
+
+  Example:
+    nix-build maintainers/nixos-tests-for-files.nix --arg files '[ nixos/modules/services/databases/postgresql.nix ]'
```
@MattSturgeon Did you try running the command before making that comment?

MattSturgeon replied:
> I've never seen a nix path that didn't start with either `.` or `/` before.

Ah, I see. Nix treats any literal containing a `/` as a path.

So while `[ foo ]` is a "`undefined variable 'foo'`" error, `[ foo/bar ]` is valid syntax.

```
$ nix-instantiate --eval -E '[ test/foo ]'
[ /home/matt/test/foo ]
$ nix-instantiate --eval -E '[ ./test/foo ]'
[ /home/matt/test/foo ]
$ nix-instantiate --eval -E '[ ./test ]'
[ /home/matt/test ]
$ nix-instantiate --eval -E '[ test ]'
error: undefined variable 'test'
       at «string»:1:3:
            1| [ test ]
             |   ^
```

TIL.

---

Therefore my suggestion would be to clarify that these are in-fact paths by using the more traditional `./` prefix:

```suggestion
    nix-build maintainers/nixos-tests-for-files.nix --arg files '[ ./nixos/modules/services/databases/postgresql.nix ]'
```

roberth replied:
If anything, I think this shows that we've done a good job at eradicating that syntax, even without a big push.
- Opened https://github.com/NixOS/nix/issues/13374

roberth committed:
maintainers/nixos-tests-for-files.nix: Use ./

This makes it clear that it's a path value literal.
We generally prefix ./ even if it's not strictly necessary, because it helps the reader.

Co-authored-by: Matt Sturgeon <matt@sturgeon.me.uk>

roberth committed:
maintainers/scripts/nixos-tests-for-files.nix: Move and clarify intent

roberth commented on maintainers/nixos-tests-for-files.nix:
```diff
@@ -0,0 +1,25 @@
+  A more detailed structure than a list is available in the `meta.tests` option value.
+
+  Example:
+    nix-build maintainers/nixos-tests-for-files.nix --arg files '[ nixos/modules/services/databases/postgresql.nix ]'
```
Changing this to `  Example usage (to be run from the Nixpkgs root):`

I think `nix-build` is generally what you want when calling this by hand, and I'm sure we can figure out that instantiate is also ok.
If we really want to make this easy, we should add a shell script wrapper with a `--dry-run` or `--eval` flag, but let's keep it simple for now.

roberth commented:
> I couldn't find a way to get back a list of the _files_ for the tests belonging to a module.

I don't think the files are what you want, especially if you can have the actual tests, which have `meta.maintainers` on them.
I guess the reason to use files would be use of CODEOWNERS?

Also relevant:
- https://github.com/NixOS/nixpkgs/pull/403839

> Could this eventually replace `all-tests.nix` and create `nixosTests` _from_ `meta.tests`? One step further, this would potentially allow a more "by-name" structure for modules, where I put the `postgresql.nix` module and all its tests into one directory...

Yes, although for nightly CI / build farm duty I think the explicit list may still be more efficient.
Also, if we were to have optional modules that need to be imported by users, we wouldn't have a complete canonical list anymore. For instance, this infrastructure doesn't find tests in profiles.

Other than that, I agree with those thoughts.
Regrouping files by domain is a generally a good thing, with the small caveat that it can be more chaotic in some cases, but I don't think that applies here.

> TLDR: I'm not sure whether the current interface of `{ nixosTests }: ...` is good.

It's an attrset argument, so that it can evolve as needed.
My intent is for it to be called through reflection, similar to what `callPackage` does, but something that suits the transition we wish.

So yes, changes are possible. My goal here is to establish something that works, and I believe it already works quite well this way.

roberth commented:
> `runTest` equally knows about the location of the test file.

https://github.com/NixOS/nixpkgs/pull/403839 knows it even better; it knows the whole imports closure, not just the top, and it can see through anonymous modules.

Sgtm, but also no problem to pivot to something else, e.g. directory based, if at some point that seems to be better.

roberth committed:
nixos: meta.tests: Handle default.nix abbreviation

roberth committed:
nixos: meta.tests: Handle recurseIntoAttrs

roberth committed:
nixos/acme: Add meta.tests

roberth committed:
maintainers/nixos-tests-for-files: Use acme in example

Not super quick, but faster than the postgresql case.

roberth committed:
nixos/meta: Support non-existent _file in meta.tests

_file can be set to a string that does not represent a file path.
This is a useful and accepted practice, so it should just work.

roberth committed:
nixos/meta: Dogfood meta.tests

wolfgangwalther reviewed (commented):
I'm not a module-magician, so can't comment much on those parts. The concept is great, my concerns about the interface resolved.

The obvious way to use this is for OfBorg to build NixOS tests. But we should also think about how to integrate this with nixpkgs-review and/or GitHub Actions CI.

Since nixpkgs-review already pulls eval results from GHA, it could also pull a list of `nixosTests` attributes to build. Alternatively, if not expensive to evaluate, nixpkgs-review could do that itself as well, ofc.

In CI, we could make use of the data, to maybe add labels. I don't think "rebuilds tests" labels with numbers make sense, because we don't really know how many tests/modules would be affected on top of those linked. But maybe something like "has required tests" or something, which gives a quick indication of "Are there any tests associated to the module, that I should watch out for?".

Maintainer pings have already been discussed.

Anything else this could be used for?

roberth commented on nixos/modules/services/databases/postgresql.nix:
```diff
@@ -879,4 +879,9 @@ in
     thoughtpolice
     danbst
   ];
+  meta.tests =
+    { nixosTests }:
+    {
+      inherit (nixosTests) postgresql;
+    };
```
Maybe the `postgresql` name is useful? I'll commit this to keep it simple though.

roberth committed:
nixos/postgresql: Simplify meta.tests

Co-authored-by: Wolfgang Walther <walther@technowledgy.de>

roberth commented:
> Alternatively, if not expensive to evaluate,

Discovery and producing the mapping should be pretty quick, based on
```
$ time nix-instantiate -A nixosTests.simple.nodes.machine.meta.maintainers --eval --strict
real	0m1.341s
```
That is after applying a trivial fix (`teams.nextcloud` -> `.members`)
Note that I'm only using `nixosTests.simple` to produce a NixOS eval conveniently. This returns the test mapping for all of NixOS.

The real cost is in the instantiation of the actual test derivations, which needs to happen locally on the nixpkgs-review machine anyway.

I don't know much about how `nixpkgs-review` does its instantiations. Ideally they'd be part of the same evaluation, and its expression could call into this PR's logic. That way `nixosTests` is memoized and no duplicate evaluation work needs to be done.

I can't estimate the cost of the `imports` graph based mapping now, but it should be cheap, reading only the modules of the test framework and the test itself, not NixOS proper. Up to a dozen?

roberth committed:
nixos/nextcloud: Fix meta.maintainers

roberth commented on nixos/modules/misc/meta.nix:
```diff
@@ -39,6 +39,73 @@ let
   };
+
+  # TODO: add to lib?
+  resolveDefaultNix = p: if lib.pathType p == "directory" then p + "/default.nix" else p;
```
I'd removed the TODO while fixing my definition, but actually that made them equivalent, so yes. This should be done, and it should be done in a separate PR.

roberth replied:
- #418824

Whichever merges first, I'll adjust the other.
